### ISSUE TITLE
Timeout expired waiting for volumes to attach or mount for pod

### ISSUE THREAD

**Issue by jbmcfarlin31** [rook/rook#2953]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
I was running through the mysql tutorial with slight variations (mainly instead of deploying mysql deploying busybox) and my pod is stuck in the following state:
```
Events:
  Type     Reason            Age                        From                    Message
  ----     ------            ----                       ----                    -------
  Warning  FailedScheduling  7m1s (x6 over 7m1s)        default-scheduler       pod has unbound immediate PersistentVolumeClaims (repeated 2 times)
  Normal   Scheduled         7m1s                       default-scheduler       Successfully assigned default/test-pod to <ip_address>
  Warning  FailedMount       <invalid> (x4 over 4m58s)  kubelet, <ip_address>  Unable to mount volumes for pod "test-pod_default(6e162e59-5b18-11e9-a689-001dd8b7007f)": timeout expired waiting for volumes to attach or mount for pod "default"/"test-pod". list of unmounted volumes=[pvc]. list of unattached volumes=[pvc default-token-xn2qp]
```

I am running kubernetes through RKE - in which I found a Rancher specific configuration for FlexVolume on the pre-reqs docs page. I made sure my kubelet had the following:
```
services:
  kubelet:
    extra_args:
      volume-plugin-dir: /usr/libexec/kubernetes/kubelet-plugins/volume/exec
    extra_binds:
      - /usr/libexec/kubernetes/kubelet-plugins/volume/exec:/usr/libexec/kubernetes/kubelet-plugins/volume/exec
```
To my understanding, the `FLEX_VOLUME_DIR` path does not need to be set in the `operator.yml` file because Rancher uses the same exact location.

I have since rebuilt and deployed the clusters (kubernetes and rook). But still see the same error. The PVC and PV's are getting created and bound successfully:
```
NAME                                                        CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM            STORAGECLASS      REASON   AGE
persistentvolume/pvc-6e147360-5b18-11e9-a689-001dd8b7007f   1Gi        RWO            Retain           Bound    default/claim2   rook-ceph-block            13m

NAME                           STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      AGE
persistentvolumeclaim/claim2   Bound    pvc-6e147360-5b18-11e9-a689-001dd8b7007f   1Gi        RWO            rook-ceph-block   13m
```

**Expected behavior:**
Pod should have had volume mounted to the PVC. 

**How to reproduce it (minimal and precise):**
Followed the tutorials for wordpress and mysql.
```
kubectl create -f operator.yml
kubectl create -f cluster.yml
kubectl create -f storageclass.yml
kubectl create -f test-pod.yml
```

Contents of `test-pod.yml`:
```
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: claim2
spec:
  storageClassName: rook-ceph-block
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
---
kind: Pod
apiVersion: v1
metadata:
  name: test-pod
spec:
  containers:
  - name: test-pod
    image: gcr.io/google_containers/busybox:1.24
    command:
    - "/bin/sh"
    args:
    - "-c"
    - "touch /mnt/SUCCESS && exit 0 || exit 1"
    volumeMounts:
    - name: pvc
      mountPath: "/mnt"
  restartPolicy: "Never"
  volumes:
  - name: pvc
    persistentVolumeClaim:
      claimName: claim2
```

**Environment**:
* OS (e.g. from /etc/os-release): 
```
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.4 LTS
Release:        16.04
Codename:       xenial
```
* Kernel (e.g. `uname -a`): `Linux  4.15.0-1036 #38~16.04.1-Ubuntu SMP Fri Dec 7 03:21:52 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`

* Cloud provider or hardware configuration: `Azure Stack`
* Rook version (use `rook version` inside of a Rook Pod): `rook: v0.9.3`
* Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"clean", BuildDate:"2018-12-03T21:04:45Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.4", GitCommit:"c27b913fddd1a6c480c229191a087698aa92f0b1", GitTreeState:"clean", BuildDate:"2019-02-28T13:30:26Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
```
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): `RKE`
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/Rook/master/toolbox.html)): 
```
  cluster:
    id:     af0da3f2-6383-4509-8a87-8a222f48bee4
    health: HEALTH_WARN
            Degraded data redundancy: 36/108 objects degraded (33.333%), 31 pgs degraded, 100 pgs undersized

  services:
    mon: 3 daemons, quorum a,c,d
    mgr: a(active)
    osd: 2 osds: 2 up, 2 in

  data:
    pools:   1 pools, 100 pgs
    objects: 36  objects, 37 MiB
    usage:   13 GiB used, 45 GiB / 58 GiB avail
    pgs:     36/108 objects degraded (33.333%)
             69 active+undersized
             31 active+undersized+degraded
```

**Comment by jbmcfarlin31** [comment]
Just to update this thread - it looks like on initial provisioning of the Pod and the PVC, there is this timeout error. However after some time passes (not sure on how long exactly) the Pod eventually attaches to the volume and spins up successfully.

I don't know if that is a bottleneck or limitation from my current environment, or if that is just a timeout bug within Rook connecting itself. It would be nice to have someone else verify.

**TL;DR**
I spun up a pod last night with a volume linking back to a claim - the PV and PVC were bound successfully, but the pod was failing because of the timeout error. I came in this morning and looked at it and the pod was now created and running with the volume mapped. I exec'd into the pod and created a file under the volume path I mapped and was able to successfully see it on the node itself.

**Comment by marcstreeter** [comment]
@jbmcfarlin31 could you look at your events to tell how long the linking took?  I am having a similar problem but I recall it not taking so long.

**Comment by jbmcfarlin31** [comment]
@marcstreeter the timeout length appears to vary depending on what application pod is trying to connect. My nginx test examples took a few minutes, and my other were less then a minute. I'm not sure if this is a bug on Rook's end, or if I need to increase the timeout thresholds in the `operator.yml` file.

**Comment by jbmcfarlin31** [comment]
Update: still having lengthy times connecting pods to pv's. The pv's and pvc's bind in no time, but the linkage for the pod to storage takes forever. Haven't figured out what the cause of this.. and am running in `rook v1.0.2`

**Comment by travisn** [comment]
@jbmcfarlin31 Do you have large PVCs? The increased mount time can be due to a recursive chown. You might check out these properties in the operator to see if it helps turning them off: https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/operator.yaml#L113-L121

**Comment by jbmcfarlin31** [comment]
@travisn my PVCs range in size. They vary between 1024Mb to 8Gi - doesn't seem to make a difference to kubernetes though, both sizes and everything in between takes anywhere from 8-45 minutes. I've seen this issue with NFS in kubernetes also.

Does anyone else see such slow binding times?

**Comment by Aisuko** [comment]
Is there have any solution? I have the same issue here. And I was deploying the helm incubator/mysqlha, pvc size "10G"

**Comment by karolgil** [comment]
I'm experiencing same issues, running rook version `v0.9.3` on EKS version `1.13.7`. For 40 and 50 GB PVCs it takes 3-4 minutes for mount to timeout. After some time then pod mounts the volumes and starts normally.

**Comment by jbmcfarlin31** [comment]
@karolgil This is the same case for me no matter what size PVC I choose - whether it is 1024Mb or 8Gi. I am using xfs for my fstype and the default options for deployment of rook/ceph.

**Comment by karolgil** [comment]
@jbmcfarlin31 I think at this point it's obvious that this is about `fsgroup` setting. I removed this option and mounted exactly the same volumes - it took little to no time to complete, about 10 seconds.

**Comment by jbmcfarlin31** [comment]
@karolgil by chance where did you remove that setting at?

**Comment by karolgil** [comment]
@jbmcfarlin31 we were setting `fsgroup` in `securityContext` field in `pod` definition, just like [explained in the docs](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-the-security-context-for-a-pod).

In our case removing:
```
securityContext:
    runAsUser: 1000
    runAsGroup: 10000
    fsGroup: 10000
```
did the trick.

**Comment by jbmcfarlin31** [comment]
@karolgil Ah, you were setting that in the `pod` spec. I don't have any `securityContext` specified in my pod specs and I still see this slow mount time. I did manage to find a setting in the `operator.yaml` file:
```
- name: ROOK_ENABLE_FSGROUP
  value: "false"
```

It was enabled by default, so I disabled it here and rebuilt the operator however I still see incredibly slow mount times.

The other thing I noticed is when I tried connecting postgres to the storageclass, every time the database did some sort of write, it would hang and timeout. I'd basically be able to create users in postgres but the second I try to create a database it would fail and crash... I can confirm this because I changed the postgres config to reference a local host path PV instead of the storageclass, and data writes were instantaneous.

This is kind of annoying considering the options for dynamic cloud native persistent storage is few and far between... would really be nice to hear from others what they have done to fix this. Otherwise I don't see how this can be running in production.

**Comment by walkerandco** [comment]
I agree, I'm having the same issue with `k3s`. I've not actually successfully bound. I've used to WP and Mysql examples from the Rook repo and I'm left with the same issue of the pod timing out when trying to attach the filesystem. The PV to PVC is immediate, but the attachment to the underlying filesystem is extremely slow.

**Comment by lz006** [comment]
same here.

we've tried several versions:
ceph
- v14.2.4
- v14.2.2
- v13.2.7
they all behaved the same.

Could anyone help us out?
-> PVC binds immediate with success:
  ```
  mysql-pv-claim   Bound    pvc-fc941ce8-e91f-4e9a-b8c1-84dcb032c026   512Mi      RWO            rook-ceph-block   7s
  ```
But pod creation stays in "ContainerCreation" which finally ends up with a timeout while trying to bind pvc/pv.

Operator prints no errors.

Can anyone help us out?

operator config:
``` yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: rook-ceph-operator
  namespace: rook-ceph
  labels:
    operator: rook
    storage-backend: ceph
spec:
  selector:
    matchLabels:
      app: rook-ceph-operator
  replicas: 1
  template:
    metadata:
      labels:
        app: rook-ceph-operator
    spec:
      serviceAccountName: rook-ceph-system
      containers:
      - name: rook-ceph-operator
        image: rook/ceph:v1.1.7
        args: ["ceph", "operator"]
        volumeMounts:
        - mountPath: /var/lib/rook
          name: rook-config
        - mountPath: /etc/ceph
          name: default-config-dir
        env:
        - name: ROOK_CURRENT_NAMESPACE_ONLY
          value: "false"
        - name: FLEXVOLUME_DIR_PATH
          value: "/var/lib/kubelet/volume-plugins"
        - name: ROOK_ALLOW_MULTIPLE_FILESYSTEMS
          value: "false"
        - name: ROOK_LOG_LEVEL
          value: "DEBUG"
        - name: ROOK_CEPH_STATUS_CHECK_INTERVAL
          value: "60s"
        - name: ROOK_MON_HEALTHCHECK_INTERVAL
          value: "45s"
        - name: ROOK_MON_OUT_TIMEOUT
          value: "600s"
        - name: ROOK_DISCOVER_DEVICES_INTERVAL
          value: "60m"
        - name: ROOK_HOSTPATH_REQUIRES_PRIVILEGED
          value: "false"
        - name: ROOK_ENABLE_SELINUX_RELABELING
          value: "false"
        - name: ROOK_ENABLE_FSGROUP
          value: "false"
        - name: ROOK_DISABLE_DEVICE_HOTPLUG
          value: "false"
        - name: ROOK_ENABLE_FLEX_DRIVER
          value: "true"
        - name: ROOK_ENABLE_DISCOVERY_DAEMON
          value: "true"
        - name: ROOK_CSI_ENABLE_CEPHFS
          value: "true"
        - name: ROOK_CSI_ENABLE_RBD
          value: "true"
        - name: ROOK_CSI_ENABLE_GRPC_METRICS
          value: "true"
        - name: NODE_NAME
          valueFrom:
            fieldRef:
              fieldPath: spec.nodeName
        # The pod name to pass with the downward API
        - name: POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        # The pod namespace to pass with the downward API
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
      volumes:
      - name: rook-config
        emptyDir: {}
      - name: default-config-dir
        emptyDir: {}
```

**Comment by lz006** [comment]
ok finally got it working by downgrading kubernetes from 1.15.3 to **1.14.6**.

Is someone able to explain why rook doesn't seem to work on k8s 1.15?

**Comment by Ehekatl** [comment]
too bad this issue marked wontfix, why is no one look into this...
@jbmcfarlin31 Have you find a way out?

**Comment by jbmcfarlin31** [comment]
@Ehekatl I really wish more investigation would go into this... as we have scrapped using it until something is changed. Unfortunately, I have not found a way to get this working or to increase mount times. We were using the CNI `calico` and noticed with other things that the MTU size was dropping packets at around the 1400 mtu size, causing significant traffic latency and interruptions. We have since switched to `weave` and a lot of those issues have fixed themselves. 

We also had a nic MTU bug in azure stack that would drop UDP packets at the 1400mtu mark as well... but that has since fixed.

When I get some free time I plan on redeploying the latest version of rook to see if the CNI potentially was our issue. I can report when I do, but I don't have a timeline for that.

**Comment by resamaraschi** [comment]
Experiencing the same issue :(

**Comment by Roarain** [comment]
Experiencing the same issue.

centos7.7 + kernel 4.20.13
k8s:1.17.1
ceph:v14.2.7

Encountered this problem many times ,waiting for auto normaly itself, have no idea about it.

**Comment by Roarain** [comment]
> Experiencing the same issue.
> 
> centos7.7 + kernel 4.20.13
> k8s:1.17.1
> ceph:v14.2.7
> 
> Encountered this problem many times ,waiting for auto normaly itself, have no idea about it.

the problem was solved by 
https://github.com/rook/rook/issues/5068

**Comment by Ehekatl** [comment]
@Roarain didn't get how the problem were solved...by reinstall ceph?

**Comment by averi** [comment]
I honestly believe this is related to the fact Kubernetes attempts to chown specific devices and storage classes behind the scenes. That would also explain why Kubernetes 1.18 introduced a flag to disable that specific behavior as per https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#configure-volume-permission-and-ownership-change-policy-for-pods. 

Recently switched a specific PVC from glusterfs into RBD (no rook, plain RBD) and noticed this exact behavior. Attaching the volume takes 2-3s then mounting it takes around 5M for a 100G volume containing loads of small files.

**Comment by vasicvuk** [comment]
We are experiencing the same issue after upgrading from Kubernetes 1.14 to 1.18.

**Comment by Aisuko** [comment]
Let's get some update from the issue

**Comment by christianhuening** [comment]
CEPH: 13.2.8
rook 0.9.3
K8s 1.15.11
Kernel. 5.4.x

We have the same issue only with a minio Statefulset, which has PVCs with 500GB size. The volumes get mounted on the Node and Flexvolume Driver is also reporting everything would be happily mounted but the Pod states a timeout. We were just updating our cluster and thus rebooting all nodes and so far the minio pods were the only pods that had this issue.

**Comment by echel0n-TR** [comment]
This issue is present for us as well, we have a erasure coding rook-ceph cluster on Kubernetes 1.26, our PVC size is 15TB with 8TB in use, it takes around 5 minutes just to mount, we looked into `fsGroupChangePolicy` but learned it does not work when the storage driver is CSI, anyone find a fix for this issue?

**Comment by echel0n-TR** [comment]
I should point out that these PVC's are RWO and not RWX, having this resolved is rather critical as the services these PVC's attached to become out of sync  and take awhile to re-sync cause of the downtime and time it takes to re-mount.

**Comment by travisn** [comment]
@echel0n-TR Would you mind opening a new issue to describe the issue in more detail? Then csi folks can take a fresh look at it.
